controller/engine_controller.go

```diff
@@ -876,6 +876,20 @@ func (m *EngineMonitor) refresh(engine *longhorn.Engine) error {
 		m.expansionBackoff.DeleteEntry(engine.Name)
 	}
 
+	defer func() {
+		if err != nil {
+			for _, replica := range replicas {
+				if replica.Spec.FailedAt == "" {
+					replica.Spec.FailedAt = util.Now()
+					if _, err := m.ds.UpdateReplica(replica); err != nil {
```

Comment:
Hmm? The engine controller updates the replica object? Shouldn't this be done in the volume controller?

Reply:
1. Restore is handled by sync agent server hence the engine process doesn't know what is the restore status and won't mark the replica as ERROR mode after the restore failure.
2. There are 3 ways to fail the replica for restore failure:
    2.1. Directly set `FailedAt` for the replica after receiving the restore status in the engine monitor. This is what we are using.
    2.2. Manually modifying the mode in `engine.Status.ReplicaModeMap` to ERROR in the engine monitor. But this will lead to the field `engine.Status.ReplicaModeMap` being inconsistent with the engine process actual status.
    2.3. Add one more gRPC call for the engine process so that the engine monitor could inform the engine process and let it mark the replicas as ERROR mode.

Reply:
Can volume controller mark the replica as failed once received the RestoreStatus update from engine object?

Reply:
It's possible. But where can we update replicas for the pre-restore check failure case?

Reply:
We can set all the replica status to failure if pre-restore check failed, or add a condition to engine for this failure.